core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java

```diff
@@ -0,0 +1,144 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Collection;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Initializes a runnable which watches files.
+ */
+class FileWatcherRunnable implements Runnable {
+    private static final Logger logger = LoggerFactory.getLogger(FileWatcherRunnable.class);
+
+    private final WatchService watchService;
+    private final Supplier<Collection<FileWatchEvent>> ctxSupplier;
+
+    /**
+     * Initializes a runnable which watches files.
+     * @param watchService the {@code WatchService} to poll events from
+     * @param ctxSupplier supplier which contains the callback invoked for each event
+     */
+    FileWatcherRunnable(WatchService watchService, Supplier<Collection<FileWatchEvent>> ctxSupplier) {
+        this.watchService = watchService;
+        this.ctxSupplier = ctxSupplier;
+    }
+
+    @Override
+    public void run() {
+        try {
+            WatchKey key;
+            while ((key = watchService.take()) != null) {
+                for (WatchEvent<?> event : key.pollEvents()) {
+                    if (event.kind().type() == Path.class) {
+                        @SuppressWarnings("unchecked")
+                        final Path watchedPath = ((Path) key.watchable())
+                                .resolve(((WatchEvent<Path>) event).context());
+
+                        if (event.kind().equals(ENTRY_DELETE)) {
+                            logger.warn("ignoring deleted file: {}", watchedPath);
+                            continue;
+                        }
+
+                        final Path realFilePath;
+                        try {
+                            realFilePath = watchedPath.toRealPath();
+                        } catch (IOException e) {
+                            logger.warn("skipping -- unable to get real path for {}", watchedPath, e);
+                            continue;
+                        }
+
+                        runCallback(realFilePath);
+                    } else {
+                        if (event.kind().equals(OVERFLOW)) {
+                            logger.debug("watch event may have been lost");
+                        } else {
+                            logger.debug("ignoring unexpected event type: {}", event.kind().name());
+                        }
+                    }
+                }
+
+                final boolean reset = key.reset();
+                if (!reset) {
+                    logger.warn("aborting file watch due to unexpected error");
+                    break;
```

Comment:
Question: Is this an error that prevents us from continue watching? What happens if we do not break the loop here?

Reply:
Good point. I wasn't able to create a test case but there was a race condition I was able to reproduce by adding a breakpoint and debugging. 
1. `watchService.take()` returns a `WatchKey`
2. `key.cancel()` is called 
3. `key.reset()` is called and returns false, breaking the loop
This would prevent losing directory watches for other "innocent" paths.

The javadoc imply that `invalid` keys don't take part in further watch events, so I agree it would be better to remove the `break` statement here